src/procedures/proc_functions.c

```diff
@@ -0,0 +1,192 @@
+(
+    ProcedureCtx *ctx,
+    const SIValue *args,
+    const char **yield
+) {
+    if(array_len((SIValue *)args) != 0) return PROCEDURE_ERR;
+
+    ProcFunctionsPrivateData *pdata = rm_malloc(sizeof(ProcFunctionsPrivateData));
+    memset(pdata, 0, sizeof(ProcFunctionsPrivateData));
+
+    // initialize an iterator to the rax that contains all functions
+    rax *functions = __aeRegisteredFuncs;
+    raxStart(&pdata->iter, functions);
+    raxSeek(&pdata->iter, "^", NULL, 0);
+    pdata->output = array_new(SIValue, 5);
+    _process_yield(pdata, yield);
+
+    ctx->privateData = pdata;
+    return PROCEDURE_OK;
+}
+
+
+// promote the rax iterator to the next function and return its name and mode.
+SIValue *Proc_FunctionsStep
+(
+    ProcedureCtx *ctx
+) {
+    ASSERT(ctx->privateData);
+    ASSERT(ctx->privateData != NULL);
+    
+    size_t bufferLen = MULTIPLE_TYPE_STRING_BUFFER_SIZE * 15;
+    char buf[bufferLen];
+
+    ProcFunctionsPrivateData *pdata = (ProcFunctionsPrivateData *)ctx->privateData;
+
+    
+    while(raxNext(&pdata->iter)){
+
+        AR_FuncDesc *func = (AR_FuncDesc*)pdata->iter.data;
+        // filter out all internal functions
+        if(!func->internal){      
+            if(pdata->yield_name && func->name ){
+                *pdata->yield_name = SI_ConstStringVal(func->name);
+            }
+            if(pdata->yield_signature){
+                // get the function signature into buf
+                SITypes_SignatureToString(func->name, func->ret_type, func->types, buf, bufferLen);
+                *pdata->yield_signature = SI_DuplicateStringVal(buf); 
+            }
+            if(pdata->yield_description && func->description){
+                *pdata->yield_description = func->description ? SI_ConstStringVal(func->description) : SI_ConstStringVal("No description available");
+            }
+            if(pdata->yield_return_description && func->ret_type){
+                // get the return value description into buf
+                SIType_ToMultipleTypeStringSimple(func->ret_type, '|', buf, bufferLen);
+                *pdata->yield_return_description = SI_ConstStringVal(buf);
+            }
+            if(pdata->yield_argument_description && func->types){
+                int arg_count = array_len(func->types);
+                if(0 < arg_count){
+                    *pdata->yield_argument_description = SI_Array(arg_count);
+                    for (int i = 0; i < arg_count; i++) {
+                        // get the argument description into buf
+                        SIType_ToMultipleTypeStringSimple(func->types[i], '|', buf, bufferLen);
+                        SIValue value = SI_ConstStringVal(buf);
+                        SIArray_Append(pdata->yield_argument_description, value);
+                    }
+                }else{
+                *pdata->yield_argument_description = SI_Array(0); 
+                }
+            }
+            return pdata->output;
+        }
+    }
+    return NULL;
```

Comment:
Add comment, iterator depleted, no more functions to yield

Reply:
Done